packages/firestore/src/util/async_queue.ts

```diff
@@ -39,25 +58,37 @@ export class AsyncQueue {
    * Adds a new operation to the queue. Returns a promise that will be resolved
    * when the promise returned by the new operation is (with its value).
    *
-   * Can optionally specify a delay to wait before queuing the operation.
+   * Can optionally specify a delay (in milliseconds) to wait before queuing the
+   * operation.
    */
   schedule<T>(op: () => Promise<T>, delay?: number): Promise<T> {
     if (this.failure) {
       fail('AsyncQueue is already failed: ' + this.failure.message);
     }
 
     if ((delay || 0) > 0) {
-      this.delayedOpCount++;
-      const deferred = new Deferred<T>();
-      setTimeout(() => {
+      this.delayedOperationsCount++;
+      const nextIndex = this.delayedOperations.length;
+      const delayedOp: DelayedOperation<T> = {
+        handle: null,
+        op: op,
+        deferred: new Deferred<T>()
+      };
+      delayedOp.handle = setTimeout(() => {
         this.scheduleInternal(() => {
-          return op().then(result => {
-            deferred.resolve(result);
+          return delayedOp.op().then(result => {
+            delayedOp.deferred.resolve(result);
           });
         });
-        this.delayedOpCount--; // decrement once it's actually queued.
+        this.delayedOperationsCount--;
+        if (this.delayedOperationsCount > 0) {
+          delayedOp.handle = null;
```

Comment:
nit but I'm not a fan of this sort of optimization. It saves an assignment in some cases, but reads much less naturally (to me at least). I'd just do:
```
delayedOp.handle = null;
if (this.delayedOperationsCount === 0) {
  this.delayedOperations = [];
}
```

If you insist on optimizing this sort of thing, you should go ahead and optimize away the decrement too!
```
if (this.delayedOperationsCount !== 1) {
  this.delayedOperationsCount--;
  delayedOp.handle = null;
} else {
  this.delayedOperations = [];
}
```
(please don't actually do that :-))

Reply:
This wasn't meant to be the newest and greatest performance optimization. Rather, it just felt naturally to do this at the time. Your version is a bit more concise, so I went with that. Thanks.